File: video-01-syntax-basics/segment-06-operators/Operators.java
// Operators.java — Segment 06: Operators
//
// Java: / truncates when both ints, Math.pow, &&, ||, !
// Python equivalent: operators.py

public class Operators {

    public static void main(String[] args) {
        System.out.println("=== Segment 06: Operators ===");

        // --- Division: int/int truncates toward zero in Java ---
        System.out.println("10 / 3  (int/int)  = " + (10 / 3));      // 3
        System.out.println("10.0 / 3 (float/)  = " + (10.0 / 3));    // 3.333...
        System.out.println("-7 / 2  (truncates) = " + (-7 / 2));      // -3 (toward zero)

        // --- Modulo ---
        System.out.println("10 % 3             = " + (10 % 3));       // 1

        // --- Exponentiation — no ** operator; use Math.pow ---
        System.out.println("Math.pow(2, 10)    = " + (int) Math.pow(2, 10));

        // --- Logical operators: symbols ---
        boolean t = true, f = false;
        System.out.println("t && f  = " + (t && f));   // Python: t and f
        System.out.println("t || f  = " + (t || f));   // Python: t or  f
        System.out.println("!t      = " + (!t));        // Python: not t

        // --- Bitwise (same symbols in both languages) ---
        System.out.println("5 & 3   = " + (5 & 3));    // 1
        System.out.println("5 | 3   = " + (5 | 3));    // 7
        System.out.println("5 ^ 3   = " + (5 ^ 3));    // 6

        // --- "in" equivalent: .contains() ---
        var roles = java.util.List.of("admin", "editor", "viewer");
        System.out.println("contains: " + roles.contains("admin"));

        // Java has no := walrus operator — assign and test are separate statements
    }
}

File: video-01-syntax-basics/segment-06-operators/operators.py
# operators.py — Segment 06: Operators
#
# Rapid-fire: Python operators that differ from Java.
# Java equivalent: Operators.java

print("=== Segment 06: Operators ===")

# --- / ALWAYS gives float in Python (not int!) ---
print(f"10 / 3  (true div)   = {10 / 3}")      # 3.3333... ← NOT 3 like Java!
print(f"10 // 3 (floor div)  = {10 // 3}")     # 3   ← use // for integer result
print(f"-7 // 2 (floors!)    = {-7 // 2}")     # -4  ← floors toward -∞ (Java gives -3)

# --- Modulo ---
print(f"10 % 3               = {10 % 3}")       # 1

# --- ** is exponentiation (no Math.pow needed) ---
print(f"2 ** 10              = {2 ** 10}")      # 1024

# --- Logical operators are English words, not symbols ---
t, f = True, False
print(f"t and f = {t and f}")   # cf. Java &&
print(f"t or  f = {t or f}")    # cf. Java ||
print(f"not t   = {not t}")     # cf. Java !

# --- Bitwise: same symbols as Java ---
print(f"5 & 3   = {5 & 3}")     # 1
print(f"5 | 3   = {5 | 3}")     # 7
print(f"5 ^ 3   = {5 ^ 3}")     # 6

# --- in operator: replaces .contains() ---
roles = ["admin", "editor", "viewer"]
print(f"'admin' in roles  = {'admin' in roles}")    # True  — cf. roles.contains("admin")
print(f"'owner' not in    = {'owner' not in roles}") # True  — not in is one operator

# --- Walrus operator := (Python 3.8+) — assign AND test in one expression ---
# Java has no equivalent; you must split into two statements
api_responses = [200, 404, 200, 500, 200]
errors = [code for code in api_responses if (category := code // 100) >= 4]
print(f"error codes       = {errors}")   # [404, 500]
# Java: must compute category separately, can't assign inside if
